### PR TITLE
[Promotion] Promotion coupon validation

### LINKED ISSUE #6308: [Promotion] Rename Coupon to PromotionCoupon
See #5384 for further reference.

### PR DESCRIPTION
**Pull request by pamil**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | yes
| BC breaks?      | yes
| Related tickets | fixes #6166, fixes #6308
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Context/Setup/PromotionContext.php [hunk 1]
```diff
@@ -130,11 +130,64 @@ public function thereIsPromotionWithCoupon($promotionName, $couponCode, $usageLi
         $promotion->setCouponBased(true);
 
         $this->promotionRepository->add($promotion);
+
         $this->sharedStorage->set('promotion', $promotion);
+        $this->sharedStorage->set('promotion_coupon', $coupon);
         $this->sharedStorage->set('coupon', $coupon);
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/PromotionContext.php]: Should be removed.

File: src/Sylius/Behat/Context/Setup/PromotionContext.php [hunk 2]
```diff
@@ -457,7 +510,7 @@ public function thePromotionWasDisabledForTheChannel(PromotionInterface $promoti
     /**
      * @Given /^the (coupon "[^"]+") was used up to its usage limit$/
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/PromotionContext.php]: Isn't this step redundant? We probably can use `(this coupon) has already reached its usage limit` implementation for that (https://github.com/Sylius/Sylius/pull/6273/files#diff-f5f3ee73d2030917831998b3bbd5cbb6R170)?

File: src/Sylius/Behat/Context/Ui/Shop/CartContext.php [hunk 3]
```diff
@@ -418,24 +418,24 @@ public function iUseCouponWithCode($couponCode)
...
-     * @Then I should be notified that promotion coupon is not valid
+     * @Then I should be notified that the coupon is invalid
      */
-    public function iShouldBeNotifiedThatPromotionCouponIsNotValid()
+    public function iShouldBeNotifiedThatCouponIsInvalid()
     {
-        $this->notificationChecker->checkNotification(
-            'Your promotion coupon is not valid.',
-            NotificationType::failure()
+        Assert::eq(
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Ui/Shop/CartContext.php]: `same`

File: src/Sylius/Bundle/CoreBundle/Resources/config/validation.xml [hunk 4]
```diff
@@ -162,6 +173,43 @@
...
+                <option name="groups">sylius</option>
+            </constraint>
+        </property>
+        <property name="defaultLocale">
+            <constraint name="NotBlank">
+                <option name="message">sylius.channel.default_locale.not_blank</option>
+                <option name="groups">sylius</option>
+            </constraint>
+        </property>
+    </class>
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/validation.xml]: Are the above changes relevant for this PR?

File: src/Sylius/Bundle/PromotionBundle/Form/Type/CouponToCodeType.php [hunk 5]
```diff
@@ -63,6 +63,34 @@ public function configureOptions(OptionsResolver $resolver)
...
+
+        return $coupon->getCode();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function reverseTransform($code)
+    {
+        if (!$code) {
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/PromotionBundle/Form/Type/CouponToCodeType.php]: Can't we use a better check? E.g. `empty($code)`?

**Comment by pjedrzejewski** [comment]
Thank you Kamil!

**Pull request merged**